docs_src/design/adr/security/0017-consul-security.md

```diff
@@ -0,0 +1,111 @@
+(Non-secure configuations continue to use Consul in
+anonymous read-write mode.)
+
+Consul provides several services for the EdgeX architecture:
+
+- Service registry (see ADR in references below)
+- Service health monitoring
+- Mutable configuration data
+
+Use of the services provided by Consul is optional on a service-by-service basis.
+Use of the registry is controlled by the `-r` or `--registry` flag provided to an EdgeX service.
+Use of mutable configuration data is controlled by the `-cp` or `--configProvider` flag provided to an EdgeX service.
+When Consul is enabled as a configuration provider,
+the `configuration.toml` is parsed into individual settings
+and seeded into the Consul key-value store on the first start of a service.
+Configuration reads and writes are then done to Consul if it is specified as the configuration provider,
+otherwise the static `configuration.toml` is used.
+Writes to the `[Writable]` section in Consul trigger per-service callbacks
+notifying the application of the changed data.
+
+Since configuration data can affect the runtime behavior of services,
+compensating controls must be introduced in order to mitigate the risks introduced
+by moving configuration from a static file into to an HTTP-accessible service with mutable state.
+
+The current practice is that Consul is exposed via unencrypted HTTP in anonymous read/write mode
+to all processes and EdgeX services running on the host machine.
+
+## Decision
+
+Consul will be configured with access control list (ACL) functionality enabled,
+and each EdgeX service will utilize a Consul access token to authenticate to Consul.
+Consul access tokens will be requested from the Vault Consul secrets engine
+(to avoid introducing additional bootstrapping secrets).
+
+DNS will be disabled via configuration as it is not used in EdgeX.
+
+**Consul Access Via API Gateway**
+
+In security enabled EdgeX, the API gateway will be configured to
+proxy the Consul service over the `/consul` path,
+using the `request-transformer` plugin
+to add the global management token to incoming requests
+via the `X-Consul-Token` HTTP header.
+Thus, ability to access remote APIs also grants the ability
+to modify Consul's key-value store.
+
+Proxying of the Consul UI is problematic and there is no current solution,
+which would involve proper balacing of the externally-visible URL,
+the path-stripping effect (or not) of the proxy,
+Consul's `ui_content_path`,
+and UI authentication
+(the `request-transfomer` does not work on the UI).
+
+
+## Consequences
+
+Full implementation of this ADR will deny Consul access to all existing Consul clients.
+To limit the impacts of the change, deployment will take place in phases:
+
+### Phase 1
+
+- Vault bootstrapper will install Vault Consul secrets engine.
+- Secretstore-setup will create a Vault token for consul secrets engine configuration.
+- Consul will be started with Consul ACLs enabled with persistent agent tokens and a default "allow" policy.
```

Comment:
Once Consul's ACL enabled, the Consul's GUI, `http://localhost:8500/ui` by default, automatically blocked whether the policy is "allow" / "deny" or not.  I've verified this locally: 
```
level=INFO ts=2021-02-16T15:21:44.481396284Z app=edgex-security-bootstrapper source=config.go:316 msg="Using local configuration from file (15 envVars overrides applied)"
level=INFO ts=2021-02-16T15:21:44.481428607Z app=edgex-security-bootstrapper source=command.go:81 msg="Security bootstrapper running listenTcp"
level=INFO ts=2021-02-16T15:21:44.482170267Z app=edgex-security-bootstrapper source=listener.go:58 msg="Security bootstrapper starts listening on tcp://edgex-core-consul:54324"
level=INFO ts=2021-02-16T15:21:45.149965354Z app=edgex-security-bootstrapper source=listener.go:67 msg="Accepted connection on edgex-core-consul:54324"
    2021-02-16T15:25:43.366Z [ERROR] agent.http: Request error: method=GET url=/v1/internal/ui/services?dc=dc1 from=172.23.0.1:53834 error="ACL not found"
    2021-02-16T15:55:29.582Z [ERROR] agent.http: Request error: method=GET url=/v1/internal/ui/services?dc=dc1 from=172.23.0.1:54050 error="ACL not found"
```
It does NOT apply to the same API calls, though.  Thus, it seems to me that Consul's UI has different mechanism to deal with ACL.

Reply:
not a real error, just a UI token cached matter. seeing as resolved.